opencog/nlp/sureal/SuRealPMCB.cc

```diff
@@ -73,32 +80,60 @@ bool SuRealPMCB::variable_match(Handle &hPat, Handle &hSoln)
     std::string sPat = _as->getName(hPat);
     std::string sPatWord = sPat.substr(0, sPat.find_first_of('@'));
     std::string sSoln = _as->getName(hSoln);
-    std::string sSolnWord = sSoln.substr(0, sSoln.find_first_of('@'));
 
     // get the WordNode associated with the word (extracted from "word@1234" convention)
     Handle hPatWordNode = _as->getHandle(WORD_NODE, sPatWord);
-    Handle hSolnWordNode = _as->getHandle(WORD_NODE, sSolnWord);
 
-    // no WordNode? reject!
-    if (hSolnWordNode == Handle::UNDEFINED)
+    // get the corresponding WordInstanceNode for hSoln
+    Handle hSolnWordInst = _as->getHandle(WORD_INSTANCE_NODE, sSoln);
+
+    // no WordInstanceNode? reject!
+    if (hSolnWordInst == Handle::UNDEFINED)
         return false;
 
-    // get the neighbor of each WordNode within LgWordCset
-    HandleSeq qPatSet =_as->getNeighbors(hPatWordNode, false, true, LG_WORD_CSET, false);
-    HandleSeq qSolnSet = _as->getNeighbors(hSolnWordNode, false, true, LG_WORD_CSET, false);
+    // get the source connectors for the solution
+    std::string scmCode = "(word-inst-get-source-conn " + SchemeSmob::to_string(hSolnWordInst) + ")";
```

Comment:
Maintainability and simplicity. You added about 100 lines of code to do something that could be done in two.  In general, less code == better code.

The added lines of code are a cut-n-paste of existing code, with only a tiny change. i.e. the function is not "orthogonal", or "not normalized" aka "not in (ortho)normal form". Failure to normalize data (eg. sql tables), proofs, and algorithms is always considered to be bad form.  Its discussed on-line and in textbooks; data strcture textbooks usually have several chapters on normalization.
 
So, for example: in this case: if there is a future change to eval_h(), whoever makes that change needs to also change eval_q() -- so the amount of code to be maintained is doubled.   The testing effort is doubled, because a bugfix in one may or may not affect the other.  Both must be tested.  Because there are two blocks of code, they will typically start to diverge, as other users make changes to one, but not the other.  Do you think this won't happen? It does: you are not the first person who has tried to make the scheme bindings better by introducing minor changes, which then decay because other bug fixes by other people (none of them me!) are not correctly propagated throughout.  Eventually something stops working, and then someone removes the duplicated code ... etc.